src/python/library/tritonclient/utils/cuda_shared_memory/__init__.py

```diff
@@ -275,140 +250,142 @@ def get_contents_as_numpy(cuda_shm_handle, datatype, shape):
+            raise CudaSharedMemoryException(
+                "The size of the shared memory region is insufficient to provide numpy array with requested size"
+            )
+        if cval_len == 0:
+            result = np.empty(shape, dtype=datatype)
         else:
-            str_offset = start_pos
-            val_buf = cast(host_addr, POINTER(c_byte * byte_size.value))[0]
-            ii = 0
-            strs = list()
-            while (ii % np.prod(shape) != 0) or (ii == 0):
-                l = struct.unpack_from("<I", val_buf, str_offset)[0]
-                str_offset += 4
-                sb = struct.unpack_from("<{}s".format(l), val_buf, str_offset)[0]
-                str_offset += l
-                strs.append(sb)
-                ii += 1
-
-            val = np.array(strs, dtype=object)
-
-            # Reshape the result to the appropriate shape.
-            result = np.reshape(val, shape)
-
-    finally:
-        c_int(_cshm_cuda_shared_memory_release_host_buffer(host_addr))
-        return result
+            val_buf = ctypes.cast(
+                host_buffer, ctypes.POINTER(ctypes.c_byte * cval_len)
+            )[0]
+            val = np.frombuffer(val_buf, dtype=datatype, offset=start_pos)
+
+            # Reshape the result to the appropriate shape. This copy is only
+            # needed as the temporary CPU buffer is cleared later by
+            # _cshm_cuda_shared_memory_release_host_buffer
+            result = np.reshape(np.copy(val), shape)
+    else:
+        str_offset = start_pos
+        val_buf = ctypes.cast(
+            host_buffer, ctypes.POINTER(ctypes.c_byte * cuda_shm_handle._byte_size)
+        )[0]
+        ii = 0
+        strs = list()
+        while (ii % np.prod(shape) != 0) or (ii == 0):
+            l = struct.unpack_from("<I", val_buf, str_offset)[0]
+            str_offset += 4
+            sb = struct.unpack_from("<{}s".format(l), val_buf, str_offset)[0]
+            str_offset += l
+            strs.append(sb)
+            ii += 1
+
+        val = np.array(strs, dtype=object)
+
+        # Reshape the result to the appropriate shape.
+        result = np.reshape(val, shape)
+
+    return result
 
 
 def set_shared_memory_region_from_dlpack(cuda_shm_handle, input_values):
     # this function basically is an implementation of 'from_dlpack'
     offset_current = 0
     for input_value in input_values:
+        dl_device = _dlpack.get_dlpack_device(input_value)
+        if dl_device is not None:
+            if not _is_device_supported(dl_device):
+                raise CudaSharedMemoryException(
+                    "DLPack device type {} is not supported".format(dl_device[0])
+                )
+            # Use stream associated with the DLPack device, otherwise, just
+            # use the stream associated with device 0.
+            if dl_device[0] == _dlpack.DLDeviceType.kDLCUDA:
+                stream = _get_or_create_global_cuda_stream(dl_device[1])
+            else:
```

Comment:
Should we check if device is kDLCPU ? or this is taken care of at this point?

Reply:
Is CPU, then there is nothing we can do, by specification we shouldn't pass non-None stream at all. Here retrieves arbitrary stream just to unify CUDA memcpy semantic below.
> On CPU and other device types without streams, only None is accepted.

Reply:
To clarify my question, suppose user wasn't careful and `input_value` is on CPU, then L333 will create stream on device 0?

Reply:
yes